### PR TITLE
Template tests for k8s backend

### PR DESCRIPTION
**Pull request by heschlie**
The goal of this is to be able to test our generated templates against some reference templates.  This means that our reference templates, and the baseline data to populate KDD and etcd will need to be updated as changes are made in other repos.

This is still buggy and the reference templates might need tweaking, as well as there are no etcd tests yet.  The steps right now for `make test-kdd` are as follows:

- Spin up etcd and k8s API server
- Launch test container with tests/ volume mounted to /tests
- Run the `test_kdd.sh` script

The `test_kdd.sh` still needs work, but the basic idea is there.

- Give us a standard `NODENAME` to use that matches our data
- Apply our manifests to populate node, tpr, and tpr data
  - Create the actual tprs and then populate some data using them
- Pull down the calicoctl repo that has the templates to test
- Compile templates via `sed` and `confd` similar to how `calico-node` does it
- Compare all of the relevant templates, show the diff for non matching templates.

I've also included a small patch to confd that didn't show until we started trying to test, we were missing an ipv6 property in the k8s backend which caused issues only with `-onetime` mode.

I'm not sure I'm creating the Dockerfile in an ideal way, and I'm not sure if we should opt to generate a new container when running `make test-*` or if we should host it and mount in the confd bin.  We can probaly clean up the yaml manifests quite a bit.

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+FROM alpine:3.6
+MAINTAINER Stephen Schlie <schlie@tigera.io>
+
+ADD bin/confd /bin/confd
+
+RUN apk --no-cache add git ca-certificates openssl
+RUN update-ca-certificates
```
Comment (tomdee) [inline Dockerfile]: Is this not done during the apk add?

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+RUN /usr/bin/wget -P /bin https://storage.googleapis.com/kubernetes-release/release/v1.6.4/bin/linux/amd64/kubectl
```
Comment (tomdee) [inline Dockerfile]: Why use `RUN` rather than just an `ADD`?

File: Makefile [hunk 3]
```diff
@@ -49,3 +50,33 @@ bin/confd: $(GO_FILES) vendor/.up-to-date
...
+test-kdd: bin/confd run-etcd-host run-k8s-apiserver
+	docker run --rm --net=host -v $(CURDIR)/tests/:/tests/ heschlie/confd-test /bin/sh -c '/tests/test_kdd.sh'
+
+## Etcd is used by the kubernetes
+run-etcd-host: stop-etcd
+	docker run --detach \
+	--net=host \
+	--name calico-etcd quay.io/coreos/etcd \
+	etcd \
+	--advertise-client-urls "http://$(LOCAL_IP_ENV):2379,http://127.0.0.1:2379,http://$(LOCAL_IP_ENV):4001,http://127.0.0.1:4001" \
```
Comment (tomdee) [inline Makefile]: Where is LOCAL_IP_ENV defined?

File: backends/k8s/client.go [hunk 4]
```diff
@@ -300,6 +300,9 @@ func populateNodeDetails(kNode *kapiv1.Node, kvps map[string]string) error {
 		kvps[nodeKey+"/network_v4"] = node.BGPIPv4Net.String()
 	}
 
+	// Some empty defaults for ipv6
```
Comment (tomdee) [inline backends/k8s/client.go]: What's this for? Is it a bugfix? Should it be in a different PR?

File: tests/test_kdd.sh [hunk 5]
```diff
@@ -0,0 +1,46 @@
+#!/bin/sh
+
+export NODENAME="kube-master"
+
+echo "Waiting for API server to come online"
+until /bin/kubectl version; do
+  sleep 1
+  done
```
Comment (tomdee) [inline tests/test_kdd.sh]: outdent

File: tests/test_kdd.sh [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+
+/bin/kubectl apply -f /tests/tprs.yaml
+/bin/kubectl apply -f /tests/nodes.yaml
+
+echo "Waiting for TPRs to apply"
+until /bin/kubectl apply -f /tests/tpr_resources.yaml; do
```
Comment (tomdee) [inline tests/test_kdd.sh]: Does this fail the first time it's run?

File: Makefile [hunk 7]
```diff
@@ -49,3 +50,33 @@ bin/confd: $(GO_FILES) vendor/.up-to-date
 	    sh -c 'go build -v -i -o $@ "github.com/kelseyhightower/confd" && \
                ( ldd bin/confd 2>&1 | grep -q "Not a valid dynamic program" || \
 	             ( echo "Error: bin/confd was not statically linked"; false ) )'
+
+.PHONY: test-kdd
+## Run template tests against KDD
+test-kdd: bin/confd run-etcd-host run-k8s-apiserver
```
Comment (tomdee) [inline Makefile]: What's your plan for naming this container and building it?

File: Dockerfile [hunk 8]
```diff
@@ -0,0 +1,30 @@
+# Copyright (c) 2015-2017 Tigera, Inc. All rights reserved.
```
Comment (tomdee) [inline Dockerfile]: I think it should be clearer that this container is just for running tests. e.g. does it really need to be in the root directory?

File: Dockerfile (hunk 1 above)
Comment (heschlie) [inline Dockerfile]: I'm not 100% sure, I'd need to test it out but most folks seemed to believe that it did not do this during the install.  I only needed this for the `wget` commands which I went ahead and replaced with `ADD` as you recommended below.

File: tests/compiled_templates/bird.cfg [hunk 9]
```diff
@@ -0,0 +1,99 @@
+# Generated by confd
```
Comment (tomdee) [inline tests/compiled_templates/bird.cfg]: These are the expected outputs right? Maybe the directory should be called expected outputs or something.

What's the plan for when these change? Could you include some comments, or documentation guiding how to fix these tests up when they start failing becuase we changed the templates in a different repo.

File: tests/nodes.yaml [hunk 10]
```diff
@@ -0,0 +1,232 @@
+apiVersion: v1
```
Comment (tomdee) [inline tests/nodes.yaml]: I think you can and should make this files _much_ smaller. As a start remove everything under `status` - it's read only anyway

File: tests/test_kdd.sh [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+echo "Getting latest confd templates from calicoctl"
+/usr/bin/git clone https://github.com/projectcalico/calicoctl.git
+/bin/ln -s /calicoctl/calico_node/filesystem/etc/calico/ /etc/calico
+
+echo "Building initial toml files"
+sed "s/NODENAME/$NODENAME/" /etc/calico/confd/templates/bird6_aggr.toml.template > /etc/calico/confd/conf.d/bird6_aggr.toml
+sed "s/NODENAME/$NODENAME/" /etc/calico/confd/templates/bird_aggr.toml.template > /etc/calico/confd/conf.d/bird_aggr.toml
+sed "s/NODENAME/$NODENAME/" /etc/calico/confd/templates/bird_ipam.toml.template > /etc/calico/confd/conf.d/bird_ipam.toml
+
+# Need to pause as running
```
Comment (tomdee) [inline tests/test_kdd.sh]: I don't understand this comment

File: backends/k8s/client.go (hunk 4 above)
Comment (heschlie) [inline backends/k8s/client.go]: I can move it to another PR if we need, but it is a small fix that was caught when running confd in `-onetime` mode, when it sees this key is missing in the `bird6.cfg` template it issues a log to `Fatal` but when run with `-watch` or `-interval` it logs to `error` so it carries on to create the rest of the templates.  This fixes that so it generates all of the templates in `-onetime` mode.

File: tests/test_kdd.sh (hunk 6 above)
Comment (heschlie) [inline tests/test_kdd.sh]: The TPRs take some time to be available for use once they are created, so we just keep trying until it takes them, which is a few seconds.

File: tests/confd_kubeconfig [hunk 12]
```diff
@@ -0,0 +1,15 @@
+apiVersion: v1
```
Comment (tomdee) [inline tests/confd_kubeconfig]: Do you actually need this file?

File: tests/compiled_templates/bird.cfg (hunk 9 above)
Comment (heschlie) [inline tests/compiled_templates/bird.cfg]: I can, I did note that in my initial PR comment that this will add the overhead of maintaining both the templates as well as the data.

File: tests/confd_kubeconfig (hunk 12 above)
Comment (heschlie) [inline tests/confd_kubeconfig]: confd does, it is not aware of first checking localhost, and we are not running as a k8s pod so it can't glean the details from the environment.

File: Makefile (hunk 7 above)
Comment (heschlie) [inline Makefile]: Looking for feedback on that, right now I just built it manually and reference what I tagged it as here for testing.  I'll add a make target to build it, but I'm not sure if we want to add that as a dependency to the test targets, or have the test targets just pull down a version of the container.  If we go that route we would need to mount in `bin/confd` to the container so we can access the latest binary.

**Comment by heschlie** [comment]
@tomdee @robbrockbank We're ready for another round.  I still need to clean up the shell scripts a bit, and move some common functionality into a `utils.sh` file.  Let me know what needs more elaboration, populating the etcd data ended up being messier than I would have liked.  I also need to grab a completed `bird6.cfg`.

Also, as @robbrockbank and I spoke, I updated the `bird.cfg compiled template to match what `confd v0.12.1-calico0.1.0` outputs due to an upstream bug, this should not impact how bird behaves, I tested this version on a couple of kdd and etcd clusters (dind and vagrant) and they passed our e2e tests.

File: Makefile [hunk 13]
```diff
@@ -49,3 +54,66 @@ bin/confd: $(GO_FILES) vendor/.up-to-date
...
+## Run all tests
+test: test-kdd test-etcd
+
+.PHONY: test-kdd
+## Run template tests against KDD
+test-kdd: bin/confd fetch-bins run-etcd-host run-k8s-apiserver
+	docker run --rm --net=host \
+	    -v $(CURDIR)/tests/:/tests/ \
+		-v $(CURDIR)/bin:/calico/bin/ \
+		heschlie/confd-test /tests/test_kdd.sh
```
Comment (tomdee) [inline Makefile]: We can't merge this pointing here...
